aws/sdk/integration-tests/s3/tests/body_size_hint.rs

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Body wrappers must pass through size_hint
+
+use aws_config::SdkConfig;
+use aws_sdk_s3::{
+    config::{Credentials, HttpClient, Region, RuntimeComponents, SharedCredentialsProvider},
+    primitives::{ByteStream, SdkBody},
+    Client,
+};
+use aws_smithy_runtime_api::{
+    client::{
+        http::{HttpConnector, HttpConnectorFuture, HttpConnectorSettings, SharedHttpConnector},
+        orchestrator::HttpRequest,
+    },
+    http::{Response, StatusCode},
+};
+use http_body::Body;
+use std::sync::{Arc, Mutex};
+
+#[derive(Clone, Debug, Default)]
+struct TestClient {
+    response_body: Arc<Mutex<Option<SdkBody>>>,
+    captured_body: Arc<Mutex<Option<SdkBody>>>,
+}
+impl HttpConnector for TestClient {
+    fn call(&self, mut request: HttpRequest) -> HttpConnectorFuture {
+        *self.captured_body.lock().unwrap() = Some(request.take_body());
+        let body = self
+            .response_body
+            .lock()
+            .unwrap()
+            .take()
+            .unwrap_or_else(SdkBody::empty);
+        HttpConnectorFuture::ready(Ok(Response::new(StatusCode::try_from(200).unwrap(), body)))
+    }
+}
+impl HttpClient for TestClient {
+    fn http_connector(
+        &self,
+        _settings: &HttpConnectorSettings,
+        _components: &RuntimeComponents,
+    ) -> SharedHttpConnector {
+        SharedHttpConnector::new(self.clone())
+    }
+}
+
+#[tokio::test]
+async fn download_body_size_hint_check() {
+    let test_body_content = b"hello";
+    let test_body = SdkBody::from(&test_body_content[..]);
+    assert_eq!(
+        Some(test_body_content.len() as u64),
+        test_body.size_hint().exact(),
+        "pre-condition check"
+    );
+
+    let http_client = TestClient {
+        response_body: Arc::new(Mutex::new(Some(test_body))),
+        ..Default::default()
+    };
+    let sdk_config = SdkConfig::builder()
```

Comment:
you can avoid the need to make a entire new struct by using `infallible_client_fn`